### ISSUE TITLE
OSBTreeCollectionManagerRemote.loadTree: java.lang.UnsupportedOperationException

### ISSUE THREAD

**Issue by jamieb22** [orientechnologies/orientdb#9677]
Andrey, any ideas what the cause of this error is. 

## Expected behavior  

No expecting an error

## Actual behavior  

2021-09-07 19:15:07 s.c.s.a.p.SearchBean [DEBUG] cant get folder with id: {a53c322ea08f3c5b}
java.lang.UnsupportedOperationException: null
        at com.orientechnologies.orient.client.remote.OSBTreeCollectionManagerRemote.loadTree(OSBTreeCollectionManagerRemote.java:68)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeCollectionManagerAbstract.loadSBTree(OSBTreeCollectionManagerAbstract.java:195)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeRidBag.loadTree(OSBTreeRidBag.java:894)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeRidBag.access$500(OSBTreeRidBag.java:76)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeRidBag$SBTreeMapEntryIterator.init(OSBTreeRidBag.java:393)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeRidBag$SBTreeMapEntryIterator.<init>(OSBTreeRidBag.java:320)
        at com.orientechnologies.orient.core.storage.ridbag.sbtree.OSBTreeRidBag.rawIterator(OSBTreeRidBag.java:469)
        at com.orientechnologies.orient.core.db.record.ridbag.ORidBag.rawIterator(ORidBag.java:206)
        at com.tinkerpop.blueprints.impls.orient.OrientVertex.getEdges(OrientVertex.java:839)
        at com.tinkerpop.blueprints.impls.orient.OrientVertex.getEdges(OrientVertex.java:747)
        at com.tinkerpop.blueprints.util.DefaultVertexQuery$DefaultVertexQueryIterable.<init>(DefaultVertexQuery.java:110)
        at com.tinkerpop.blueprints.util.DefaultVertexQuery.edges(DefaultVertexQuery.java:80)
        at com.tinkerpop.pipes.transform.VertexQueryPipe.processNextStart(VertexQueryPipe.java:113)
        at com.tinkerpop.pipes.transform.VertexQueryPipe.processNextStart(VertexQueryPipe.java:19)
        at com.tinkerpop.pipes.AbstractPipe.next(AbstractPipe.java:89)
        at com.tinkerpop.pipes.filter.FilterFunctionPipe.processNextStart(FilterFunctionPipe.java:23)
        at com.tinkerpop.pipes.AbstractPipe.hasNext(AbstractPipe.java:98)
        at com.tinkerpop.pipes.util.Pipeline.hasNext(Pipeline.java:105)
        at com.stimulus.archiva.database.blueprints.BluePrintVersionedItem.iterateVersions(BluePrintVersionedItem.java:210)
        at com.stimulus.archiva.database.blueprints.BluePrintVersionedItem.getMostRecentVersions(BluePrintVersionedItem.java:175)
        at com.stimulus.archiva.presentation.SearchBean.listFolderItems(SearchBean.java:1870)
        at com.stimulus.archiva.presentation.SearchBean.writeTableData(SearchBean.java:2396)
        at com.stimulus.archiva.presentation.FetchTableData.execute(FetchTableData.java:33)
        at org.apache.struts.action.RequestProcessor.processActionPerform(RequestProcessor.java:425)
        at org.apache.struts.action.RequestProcessor.process(RequestProcessor.java:229)
        at org.apache.struts.action.ActionServlet.process(ActionServlet.java:1975)
        at org.apache.struts.action.ActionServlet.doGet(ActionServlet.java:455)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at jespa.http.HttpSecurityService.doFilter(HttpSecurityService.java:1590)
        at com.stimulus.archiva.security.MailArchivaSecurityFilter.doFilter(MailArchivaSecurityFilter.java:220)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at com.stimulus.util.UTF8Filter.doFilter(UTF8Filter.java:44)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:610)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:516)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1086)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:659)
        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:223)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1558)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1515)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)

**Comment by jamieb22** [comment]
I just read this comment: https://githubmemory.com/repo/orientechnologies/orientdb/issues/9661. Apparently, OrientDB will upgrade the DB to use a new tree based representation of edges. This automatic upgrade process doesn't seem to be happening with this database. Unfortunately, this db cannot be upgraded to Orient DB to V3.2 due to the presence of show stopper issue https://github.com/orientechnologies/orientdb/issues/9568

**Comment by jamieb22** [comment]
Will Orient DB 3.1.7 eventually upgrade the DB to support fetching vertices by relation or do I need to somehow implement a workaround?

**Comment by andrii0lomakin** [comment]
Hi, @jamieb22 that will be fixed for sure.

**Comment by tglman** [comment]
Hi,

The support for edges iteration should be already there, the problem I see here is that is trying to use `OSBTreeRidBag` to iterate the relationships instead of using `ORemoteTreeRidBag`, would be useful if you could track down where the instance  of `OSBTreeRidBag` is created in your flow.

Regards

**Comment by jamieb22** [comment]
We are using Blueprints 2.6.0 to iterate the edges. if what you say is true, does it then follow that Orient DB integration with Blueprints hasn't yet been updated to using ORemoteTreeRidBag?

```
[INFO] +- com.orientechnologies:orientdb-graphdb:jar:3.1.7:compile
[INFO] |  +- com.orientechnologies:orientdb-server:jar:3.1.7:compile
[INFO] |  |  \- com.orientechnologies:orientdb-client:jar:3.1.7:compile
[INFO] |  +- com.orientechnologies:orientdb-tools:jar:3.1.7:compile
[INFO] |  +- com.orientechnologies:orientdb-core:jar:3.1.7:compile
[INFO] |  |  +- com.github.jnr:jnr-posix:jar:3.0.50:compile
[INFO] |  |  |  +- com.github.jnr:jnr-ffi:jar:2.1.10:compile
[INFO] |  |  |  |  +- com.github.jnr:jffi:jar:1.2.19:compile
[INFO] |  |  |  |  +- com.github.jnr:jffi:jar:native:1.2.19:runtime
[INFO] |  |  |  |  +- org.ow2.asm:asm-analysis:jar:7.1:compile
[INFO] |  |  |  |  +- org.ow2.asm:asm-util:jar:7.1:compile
[INFO] |  |  |  |  +- com.github.jnr:jnr-a64asm:jar:1.0.0:compile
[INFO] |  |  |  |  \- com.github.jnr:jnr-x86asm:jar:1.0.2:compile
[INFO] |  |  |  \- com.github.jnr:jnr-constants:jar:0.9.12:compile
[INFO] |  |  +- com.googlecode.concurrentlinkedhashmap:concurrentlinkedhashmap-lru:jar:1.4.2:compile
[INFO] |  |  \- org.lz4:lz4-java:jar:1.4.0:compile
[INFO] |  +- com.tinkerpop.blueprints:blueprints-core:jar:2.6.0:compile
[INFO] |  |  +- org.codehaus.jettison:jettison:jar:1.3.3:compile
[INFO] |  |  |  \- stax:stax-api:jar:1.0.1:compile
[INFO] |  |  \- com.carrotsearch:hppc:jar:0.6.0:compile
[INFO] |  +- org.codehaus.groovy:groovy:jar:2.5.8:compile
[INFO] |  \- org.codehaus.groovy:groovy-jsr223:jar:2.5.8:compile
```

**Comment by jamieb22** [comment]
I don't think we are doing anything too customized here. We are creating a graph factory and using the api as follows. The code worked fine in 3.0.x both when connecting to a local database and a remote database. In V3.1.x it throws an UnsupportedException when using a remote database, but local database works fine.

  ```
      graphFactory = new OrientGraphFactory(settings.getUrl(), settings.getLogin(), settings.getPassword(), false);

    @Override
    public void iterateVersions(Found found, Relationship relationship, Date date) throws InterruptedException {
        Graph graph = graphFactory.getTx().getGraph();
        try {
            GremlinPipeline<Vertex, ?> pipe = new GremlinPipeline(vertex, true);

            pipe.start(vertex).outE(relationship.toString().toLowerCase(Locale.ENGLISH))
                    .filter(edge -> {
                        ItemVersion version = new BluePrintItemVersion(database, edge.getVertex(Direction.IN), edge);
                        if (version.getCreationDate().before(date)) {
                            if (version.isDeleted() && version.getDeletedDate().after(date) || !version.isDeleted()) {
                                return Boolean.TRUE;
                            }
                        }
                        return Boolean.FALSE;
                    });

            while (pipe.hasNext()) {
                Edge edge = (Edge) pipe.next();
                Vertex vertex = edge.getVertex(Direction.IN);
                found.found(new BluePrintItemVersion(database, vertex, edge));
            }
        } finally {
            graph.shutdown();
        }
    }


```

**Comment by tglman** [comment]
Hi,
This should have been fixed in 3.1.13

Regards,

**Comment by andrii0lomakin** [comment]
Closing the issue because of the fix provided.
